### ISSUE TITLE
with recursive CTE works on SQL Lab, but not on making chart with it

### ISSUE THREAD

**Issue by brian-bh** [apache/superset#21715]
#### How to reproduce the bug

I am using databases on Amazon Redshift (and PostgreSQL just in case).
For the purpose of ordering data to date, I was using with recursive CTE in many queries.
The problem is: recursive CTE works well in SQL Lab, but not in making Charts.

### Expected results

Making chart also works well with recursive CTE.

### Actual results

Chart query gives me an error: "Error: Recursive CTE in subquery are not supported."

#### Screenshots

(This is a partial query from my company, so I am masking some not important values and schema name)

![with_recursive_sql_lab](https://user-images.githubusercontent.com/46946734/194273543-b8978225-fea6-4910-b891-6b6b7a1ec168.PNG)

![with_recursive_on_chart](https://user-images.githubusercontent.com/46946734/194273123-3e8bc871-6898-409a-bc93-8086230238b4.PNG)

### Environment

- browser type and version: Chromium based browsers, spotted on Google Chrome
- superset version: `superset version`: 2.0.0
- on docker non-dev composed settings;

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [V] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [V] I have reproduced the issue with at least the latest released version of superset.
- [V] I have checked the issue tracker for the same issue and I haven't found one similar.

## Python Log
```
Query SELECT ym AS ym
FROM
  (with recursive x(ym) as
     (select to_char(min(date(ri.(MASKED)_date)), 'yyyy-mm') as ym
      from (MASKED) ri
      where ri.user_id in ((MASKED))
      union all select to_char(dateadd(month, 1, to_date(ym, 'yyyy-mm')), 'yyyy-mm')
      from x
      where to_date(ym, 'yyyy-mm') < to_date(dateadd(month, 0, current_date), 'yyyy-mm') ) select x.ym
   from x) AS virtual_table
LIMIT 1000 on schema pfproduct failed
Traceback (most recent call last):
  File "/app/superset/connectors/sqla/models.py", line 1924, in query
    df = self.database.get_df(sql, self.schema, mutator=assign_column_label)
  File "/app/superset/models/core.py", line 458, in get_df
    self.db_engine_spec.execute(cursor, sqls[-1])
  File "/app/superset/db_engine_specs/base.py", line 1329, in execute
    raise cls.get_dbapi_mapped_exception(ex)
  File "/app/superset/db_engine_specs/base.py", line 1327, in execute
    cursor.execute(query)
psycopg2.errors.FeatureNotSupported: Recursive CTE in subquery are not supported.
```

**Comment by mayurnewase** [comment]
Looks like redshift doesn't support recursive CTE in subquery, and this happens because chart exploration needs to wrap raw sql queries in a subquery and have control over it for columns and metrics.

I tested below with postgres works fine even on plotting chat
`WITH RECURSIVE t(n) AS (
    VALUES (1)
  UNION ALL
    SELECT n+1 FROM t WHERE n < 2
)
SELECT sum(n) FROM t;`

can you test this query in **sqllab**

`SELECT sum
FROM
  (WITH RECURSIVE t(n) AS
     (
      VALUES (1)
      UNION ALL SELECT n+1
      FROM t
      WHERE n < 2 ) SELECT sum(n)
   FROM t) AS virtual_table
LIMIT 1000;`

**Comment by brian-bh** [comment]
@mayurnewase thanks for your reply.
I tried as you say so (modified VALUES to SELECT because it gives me an error that there's no function name VALUES);
and voila, it gives me same error that I reported in SQLLab.
and the error is not only in redshift but also in postgresql, and I think that's natural in psycopg2.
but the thing is, you said that this query
`WITH RECURSIVE t(n) AS ( VALUES (1) UNION ALL SELECT n+1 FROM t WHERE n < 2 ) SELECT sum(n) FROM t;`
succeed in making charts, but I get the same error as I reported.
![image](https://user-images.githubusercontent.com/46946734/194996519-1c511654-c772-4cfa-b0ff-d09a94c5094b.png)

So, I think in making Charts, the whole query becomes subquery that named virtual_table, and that's the problem.
In this case, there's no option to make charts from queries that has recursive CTE and so?
